segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,308 @@
+
+        public EventProcessorImpl(@NonNull String name, @NonNull DirectSegmentAccess segment,
+                                  @NonNull Function<List<BufferView>, CompletableFuture<Void>> handler,
+                                  @NonNull EventProcessorConfig config) {
+            super(name, handler, config);
+            this.segment = segment;
+            this.outstandingBytes = new AtomicLong(0);
+            this.lastIterationLatency = new AtomicLong(0);
+            this.metrics = new SegmentStoreMetrics.EventProcessor(name, containerId);
+        }
+
+        @Override
+        public CompletableFuture<Long> add(@NonNull BufferView event, Duration timeout) {
+            Preconditions.checkArgument(event.getLength() > 0);
+            Preconditions.checkState(outstandingBytes.get() < MAX_OUTSTANDING_BYTES,
+                    "Too many outstanding events for {}.", this.getName());
+
+            // Appends may be variable in size. We encode each event as [EVENT_LENGTH + EVENT_DATA]. This is needed to
+            // understand the boundaries of events in the absence of an actual notion of event.
+            return segment.append(BufferView.builder().add(intToByteArraySegment(event.getLength())).add(event).build(), null, timeout)
+                    .thenApply(offset -> outstandingBytes.addAndGet(event.getLength() + Integer.BYTES));
+        }
+
+        @Override
+        public void close() {
+            metrics.close();
+        }
+    }
+
+    @Override
+    public void close() {
+        log.info("Closing ContainerEventProcessor service.");
+        closed.set(true);
+    }
+
+    //endregion
+
+    //region AbstractService Implementation
+
+    @Override
+    protected void shutDown() throws Exception {
+        log.info("Shutting down ContainerEventProcessor service.");
+        eventProcessorMap.forEach((k, v) -> v.close());
+        eventProcessorMap.clear();
+        log.info("Shut down of ContainerEventProcessor service complete.");
+    }
+
+    @Override
+    protected void startUp() throws Exception {
+        // A ContainerEventProcessor iteration is made of the following stages (for each EventProcessor registered):
+        // 1. Read the LAST_PROCESSED_EVENT_OFFSET attribute to know the starting read offset and read at most getMaxItemsAtOnce events.
+        // 2. The collected results are passed to the handler function in EventProcessor.
+        // 3. If the handler function has been successfully executed, truncate the internal Segment of the EventProcessor
+        // according to the last successfully processed event. If an error occurs, throw and re-try.
+        // 4. Report metrics for each EventProcessor registered and wait for ITERATION_SLEEP until the next iteration.
+        log.info("Starting ContainerEventProcessor service.");
+        Futures.loop(
+                () -> !closed.get(),
+                () -> Futures.delayedFuture(ITERATION_SLEEP, this.executor)
```

Comment:
You need to be able to interrupt this delayed future when you shut down. Is there an overload you can use? Otherwise your shutdown may take longer than expected.

Reply:
Replaced the use of `closed` by `CancellationToken`, which it looks to be cancelling subsequent futures in the chain, right?